src/annoymodule.cc

```diff
@@ -236,30 +236,52 @@ py_an_save(py_annoy *self, PyObject *args, PyObject *kwargs) {
 
 PyObject*
 get_nns_to_python(const vector<int32_t>& result, const vector<float>& distances, int include_distances) {
-  PyObject* l = PyList_New(result.size());
-  if (l == NULL) {
-    return NULL;
+  PyObject* l = NULL;
+  PyObject* d = NULL;
+  PyObject* t = NULL;
+  PyObject* res = NULL;
+  PyObject* dist = NULL;
+
+  if ((l = PyList_New(result.size())) == NULL) {
+    goto error;
+  }
+  for (size_t i = 0; i < result.size(); i++) {
+    if ((res = PyInt_FromLong(result[i])) == NULL) {
+      goto error;
+    }
+    if (PyList_SetItem(l, i, res) < 0) {
+      goto error;
+    }
   }
-  for (size_t i = 0; i < result.size(); i++)
-    PyList_SetItem(l, i, PyInt_FromLong(result[i]));
   if (!include_distances)
     return l;
 
-  PyObject* d = PyList_New(distances.size());
-  if (d == NULL) {
-    return NULL;
+  if ((d = PyList_New(distances.size())) == NULL) {
+    goto error;
   }
-  for (size_t i = 0; i < distances.size(); i++)
-    PyList_SetItem(d, i, PyFloat_FromDouble(distances[i]));
 
-  PyObject* t = PyTuple_New(2);
-  if (t == NULL) {
-    return NULL;
+  for (size_t i = 0; i < distances.size(); i++) {
+    if ((dist = PyFloat_FromDouble(distances[i])) == NULL) {
+      goto error;
+    }
+    if (PyList_SetItem(d, i, dist) < 0) {
+      goto error;
+    }
+  }
+
+  if ((t = PyTuple_Pack(2, l, d)) == NULL) {
+    goto error;
   }
-  PyTuple_SetItem(t, 0, l);
-  PyTuple_SetItem(t, 1, d);
 
   return t;
+
+  error:
+    Py_XDECREF(l);
+    Py_XDECREF(d);
+    Py_XDECREF(t);
+    Py_XDECREF(res);
```

Comment:
This is actually incorrect because PyList_SetItem steals a reference to `res`. So if we get here because `d = PyList_New` failed, we'll decref the last element of the `l` list twice: once when the list itself gets deleted, and once from this line. I believe setting to NULL at the end of the loop body will fix this.

Similar reasoning applies to the `dist` variable in the second for loop.

Reply:
fixed. Thanks for the feedback!

Reply:
Looks like you fixed only the second instance (dist), not the first one (res)

Reply:
I think I've fixed this now with my latest commit

Reply:
nvm, I messed it up again 🤦

Reply:
Because of the way we call PyList_SetItem, it will always succeed. We don't need to check the error response.

I _think_ I've actually resolved the double decref's.